content/en/apps/guides/forms/rdtoolkit-integration.md

```diff
@@ -0,0 +1,219 @@
+keywords: RDToolkit RDT
+
+---
+
+_Introduced in v3.12.0_
+
+CHT has an integration with [Rapid Diagnostic Toolkit (RDToolkit)](https://github.com/dimagi/rd-toolkit) which is an Android app developed by Dimagi that facilitates the use of Rapid Diagnostic Test (RDT) in global health care. 
+The integration is only available in the CHT Android app and consists of two Enketo widgets (rdtoolkit-provision and rdtoolkit-capture) that are used in XForms to provision RDT and capture results.
+
+## Create the XForms
+It is required to create one form for provisioning RDT and one form for capturing RDT results, see [the documentation](https://docs.communityhealthtoolkit.org/apps/tutorials/app-forms/) for more details about app forms.
+
+### Provisioning RDT form
+The following label is required in the `appearance` tag of `data` group:
+* `rdtoolkit-provision`: To attach the Enketo widget that connects with the RDToolkit app to provision RDT.
+
+This form has the following mandatory fields:
+* `patient_uuid` | Text
+    * Patient's unique ID in the system.
+    * Defined in the root level.
+    * Auto filled when using the contact lookup field.
+* `patient_id` | Text
+    * Patient's representative ID, for example, can be the ID in the health facility's records.
+    * Defined in the root level.
+    * Auto filled when using the contact lookup field.
+* `patient_name` | Text
+    * Patient's first name and / or last name
+    * Defined in the root level.
+    * Auto filled when using the contact lookup field.
+* `rdtoolkit_session_id` | Text
+    * The unique ID for the RD Test.
+    * Defined in the `data` group level.
+    * Auto filled by CHT integration.
+* `rdtoolkit_filter` | Text
+    * The group of RDT that users can select when provisioning RDT by using RDToolkit app.
+    * Defined in the `data` group level.
+    * The available options can be found in the [RDToolkit app source code](https://github.com/dimagi/rd-toolkit).
+    * Assign a group name to enable all the RDT associated with the group, example: `mal_pf`.
+    * Assign individual RDT names to limit the options available for the user, separated by one space, example: `sd_bioline_mal_pf_pv carestart_mal_pf_pv`.
+
+Optional fields:
+* `rdtoolkit_api_url` | Text
+    * API URL that will receive the RDT results, sent by RDToolkit app via `POST`. See [RDToolkit app source code](https://github.com/dimagi/rd-toolkit) for more details.
+    * Defined in the `data` group level.
+* `rdtoolkit_state` | Text
+    * The RDT status.
+    * Defined in the `data` group level.
+    * Auto filled by the CHT integration.
+* `rdtoolkit_time_started` | Text
+    * The date and time when the RDT started.
+    * Defined in the `data` group level.
+    * Auto filled by the CHT integration.
+* `rdtoolkit_time_resolved` | Text
+    * The date and time when the RDT result is ready to capture.
+    * Defined in the `data` group level.
+    * Auto filled by the CHT integration.
```

Comment:
Seems like a lot of fields that need to be manually added to the XLS Form. It would have been nice for the widget to automatically add the required fields. Are they explicitly set so that the inputs can be set, and perhaps so that the output can be used in the form?

Also, if these fields persist they may be better structured as a table with columns: `Field name`, `XLSForm type`  `Description`, `Parameter type` (eg input or output), and `required`.

Reply:
Right, I'm going to switch this documentation to a table.

> It would have been nice for the widget to automatically add the required fields.

The downside I see about automatically added fields is that it isn't obvious they exits or what is group/level or how are they called when looking at the form definition (fastest way to know the actual form fields), then app devs need to take the extra step of opening the documentation or finding the form in the db, this happens "n" times. While defining the fields is a less frequent task that happens 1 time.

Reply:
Ya, I see what you are saying, and that makes sense for the return values. It is in fact how ODK XLSForms have [intents that return multiple values](https://docs.getodk.org/launch-apps-from-collect/#external-apps-to-populate-multiple-fields). 

For the input parameters, you can see the way ODK handles that in their [external intent notation](https://docs.getodk.org/launch-apps-from-collect/#specifying-a-uri-as-intent-data). In the `body::intent` column they pass the URI and then the parameters with the XPath notation (eg `uuid=/myform/group/patient_uuid, ...`) or XLSForm variable notation (eg `uuid=${patient_uuid}, ...`). Using this notation makes it clean and more expected to XLSForm developers, and with the input parameters included like this we don't require unnecessary data fields either. There are some tricky aspects of how you can pass data to the XForm from XLSForm (eg like `body::intent` as used in external intents, and `instance::type`). You might find [this documentation on prefixes that add data to XForm attributes](https://xlsform.org/en/#advanced-use-and-extensibility) helpful for that. If you have a proposed format I am happy to review together and discuss before you reimplement it.

Reply:
I'm learning this form tech on the way 😅 and not sure if I'm understanding the idea 🤔 

If using the ODK: 
As I can read the [external intent in general](https://docs.getodk.org/launch-apps-from-collect/#specifying-a-uri-as-intent-data) are available through their Android App (to be able to trigger intents) that would require users to install this extra app [ODK Collect](https://docs.getodk.org/collect-install) (+ cht-android, + rdtookit app) which will change the current flow too.

If not using that, then:
I think it's possible to create a custom column with the parameters, instead of typing these in rows then they will type them in 1 cell like this: 

[   myColumn  ]
[  rdtoolkit?filter="mal_pv"&api="https://example.com"   ]
Basically just `rdtoolkit_filter` and the `rdtoolkit_api_url` because the rest are autofilled, this solution is probably is new for us 🤔.

Because fields like `patient_uuid` still needs to be there since that's coming from the `contact` source control and managed somewhere else in cht 🤔 gotta test. Regarding the `patient_id`, `patient_name` and `rdtoolkit_session_id`, I think those should stay as fields, because they probably want to display it as read only to cross check with the dashboards or the rdtoolkit app. 

I try to imagine myself filling the form as app dev to feel the difference between solutions, however this is something I would suggest to test with users because of effort/time in development/testing/review, we'll probably get insights about other pieces of this feature too. Otherwise we can go ahead with the second solution (custom column) 🙂

Reply:
Sorry for the confusion! No need to use ODK Collect itself, just imitate it's technique. Most forms that work in ODK Collect would work in the CHT, but the reverse is not always true because of CHT-specific widgets (eg the contact-selector).

In terms of other input fields, they might be in the form, but not necessarily in the same group. For consistency with external intents, I'd suggest a `body::intent` column have the following:
`org.rdtoolkit.provision(filter="mal_pv", api="https://example.com", uuid=/form/inputs/patient_uuid, id=${patient_id}, name=${patient_name})`

If you want to use an `appearance` of `rdtoolkit` that would ensure that the rdtoolkit widget is used, which may make it more future proof if a generic external intent widget is added _later_ (which would be a nice feature, and help us be compatible with ODK XForms).

That makes it have the following **inputs** passed in the _parameter list_, all of which we'd generally use but in theory could be optional:
- `id` (required, only to differentiate tests)
- `name` (required, only to differentiate tests)
- `uuid` (how is this used?)
- `filter`
- `api`

And the **return** values as **fields within the same group**, all of which could be optional:
- `rdtoolkit_session_id`
- `rdtoolkit_state`
- `rdtoolkit_time_started`
- `rdtoolkit_time_resolved`

Something similar would also need to be done for the capture side.

Reply:
Basically, support both form fields and "query string".  I'm going to do some experiments, it might take some time as I research

Reply:
Right, and to clarify _inputs_ would only be via "query string", and _outputs_ would be form fields.

Reply:
From the UX/UI perspective, we still need the patient data (inputs) as fields because that's how the CHW know who they are provisioning RDTs: 
<img width="450" alt="Screen Shot 2021-06-14 at 11 12 28 am" src="https://user-images.githubusercontent.com/66472237/121913361-890d4b80-cd5b-11eb-9ab6-a2e4b16feef8.png"> 

Patient's uuid is needed for the contact source field association the patient, example: if there's a value it won't ask for the contact, if it's empty then the form ask for the contact (that's how it's working already in cht) it saves one step to the CHW:
<img width="450" alt="Screen Shot 2021-06-14 at 10 02 09 pm" src="https://user-images.githubusercontent.com/66472237/121914114-2d8f8d80-cd5c-11eb-9928-dae76bda91e2.png">

What I'm going to do is to leave open the possibility of receiving the contact information by query string, if this information is missing then I'll take it from the fields. I'll explain both in the docs, but the example probably more oriented to having the patient as fields since it's probably the most common scenario.  🤔

Reply:
Similar to [above](https://github.com/medic/cht-docs/pull/510#discussion_r651094219), we should be following the XLSForm convention here and having the inputs in the query string as the main way to use this widget. Generally those patient fields are already at the top level, and don't need to be repeated, which causes duplication in the resulting data as well. Being able to have the fields in the group is a nice _fallback_ if fields are missing in the query string though, but may be adding unnecessary complexity to code and documentation. Feel free to follow up offline if further clarification is needed.